src/plugins/WebOfTrust/util/SynchronousDelayedBackgroundJob.java

```diff
@@ -0,0 +1,231 @@
+     * 1. Waits for the deadline to expire
+     * 2. Waits until the previous job is done, if any
+     * 3. Resets {@link #nextRunDeadline} / {@link #nextRunScheduled}
+     * 4. Runs the background job
+     * 5. Increments {@link #runCount}
+     * 6. Notifies all threads waiting upon its completion
+     * Caller must ensure synchronization on {@code this}.
+     */
+    private void startNewJobThread() {
+        Thread jobThread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                synchronized(SynchronousDelayedBackgroundJob.this) {
+                    long delay;
+                    // 1. Wait for the deadline to expire
+                    while ((delay = nextRunDeadline - System.currentTimeMillis()) > 0) {
+                        try {
+                            SynchronousDelayedBackgroundJob.this.wait(delay);
+                        } catch (InterruptedException e) {
+                            // Ignore. We can't really return here, since there is at
+                            // least one thread waiting for this thread to complete
+                            // (possibly including side effects caused by its run).
+                        }
+                    }
+                    // 2. Wait until the previous job is done, if any
+                    while (runningJobThread != null) {
+                        try {
+                            SynchronousDelayedBackgroundJob.this.wait();
+                        } catch (InterruptedException e) {
+                            // Ignore. This thread is not even visible from the
+                            // outside at this point, and we will be notified once the
+                            // currently running job is finished. We can't really return
+                            // here, since there is at least one thread waiting for this
+                            // thread to complete (possibly including side effects
+                            // caused by its run).
+                        }
+                    }
+                    // 3. Reset the nextRun variables
+                    nextRunDeadline = NO_DEADLINE;
+                    nextRunScheduled = false;
+                    // From here on, this thread is visible from the outside world,
+                    // and interruption should be handled correctly.
+                    runningJobThread = Thread.currentThread();
+                    // We cannot return here if we are terminating, since that would violate the
+                    // contract specified by {@link #triggerExecution()}. Instead, interrupt the
+                    // thread so that the job may terminate early.
+                    if (isTerminating) {
+                        runningJobThread.interrupt();
+                    }
+                }
+                try {
+                    // 4. Run the background job
+                    job.run();
+                } finally {
```

Comment:
It's not up to a utility class to decide on how to handle uncaught exceptions in external code). If we need to log it like you suggest, Fred should provide this default behaviour by setting the appropriate default uncaught exception handler on `Thread` (`Thread.setDefaultUncaughtExceptionHandler`). Otherwise, IMHO it should be up to the application code to log its own errors. In this class we only need to care about internal integrity, i.e. we should be able to proceed even in the presence of uncaught exceptions in external code. AFAIK the latter works just fine.